File: methods/BackwardEuler.py
import numpy as np 
from scipy.optimize import fsolve

def solve_backward_euler(f, t0, t_end ,y0, h ):
    # f is the differential equation (dy/dt = f(y,t)
    #t0 is the start time 
    #t_end is the end time 
    #we create an array of values of t between t0 and t_end and compute values at those points 
    #h is the varying step size 
    #y0 is the initial condition

    if h <= 0:
        raise ValueError("Step size must be positive.")

    N = int((t_end - t0)/h) # N is the no of time steps 

    #array of time values 
    t_values = np.linspace(t0 , t_end, N+1)
    

    y_values = np.zeros_like(t_values)
    y_values[0] = y0

    for i in range(N):
        y_n = y_values[i]
        t_n = t_values[i]
        #define equation to solve for y_n+1

        def equation(y_next):
            #update equation for backward euler 
            return y_next - y_n - h* f(t_n + h , y_next)

        y_next = fsolve(equation , y_n)[0]

        y_values[i+1] = y_next

    return t_values , y_values 

File: NmethodsCodes/BackwardEuler/BackwardEuler.c
#include <stdio.h>
#include <math.h>
#include <stdlib.h>

#define TOL 1e-6  

// Define the differential equation
double f(double t, double i, double V0, double R, double L) {
    return (V0 - R * i) / L;
}

// Newton method to compute the next step 
double solve(double t_next, double i_n, double h, double V0, double R, double L) {
    double i_next = i_n; // Initial guess
    double F, dF;

    for (int iter = 0; iter < 100; iter++) { // Max 100 iterations to ensure convergence
        F = i_next - i_n - h * (V0 - R * i_next) / L;
        dF = 1 + h * (R / L); // Corrected derivative

        if (fabs(F) < TOL) 
            break; // Converged

        i_next -= F / dF; // Update using Newton's method
    }
    
    return i_next;
}

// BackwardEuler function
void Backward_euler(double (*f)(double, double, double, double, double), 
                    double t0, double t_end, double i0, double h, 
                    double V0, double R, double L, double *t_val, double *i_val) {
    
    if (h <= 0) {
        printf("Step size must be positive.\n");
        return;
    }

    int N = (int)((t_end - t0) / h); // Define number of time steps

    // Initialize values
    t_val[0] = t0;
    i_val[0] = i0;

    // Compute points in loop
    for (int i = 0; i < N; i++) {
        t_val[i + 1] = t_val[i] + h; // Update time step

        // Determine V0 value based on time
        double V_input = (int)(t_val[i + 1]) % 4 < 2 ? V0 : 0; 

        // Solve using Newton's method with V_input
        i_val[i + 1] = solve(t_val[i + 1], i_val[i], h, V_input, R, L); 
    }
}
